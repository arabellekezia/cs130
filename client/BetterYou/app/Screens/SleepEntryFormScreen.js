import React, { useState } from "react";

import { StyleSheet, View } from "react-native";

import Screen from "../components/Screen";
import AppDateTimePicker from "../components/AppDateTimePicker";
import AppText from "../components/AppText";
import TextButton from "../components/TextButton";
import colors from "../config/colors";
import ErrorMessage from "../components/ErrorMessage";
import { ScrollView, Switch } from "react-native-gesture-handler";

import SleepService from "../services/SleepService";

function isValidInput(startDate, endDate, setError) {
  const validStartDate = isValidDate(startDate);
  const validEndDate =
    isValidDate(endDate) && endDate.getTime() > startDate.getTime();
  setError({ startDate: !validStartDate, endDate: !validEndDate });
  return validStartDate && validEndDate;
}

function isValidDate(date) {
  return date instanceof Date && !isNaN(date);
}

function SleepEntryFormScreen({ navigation, route }) {
  const initStartDate = route.params ? new Date(route.params.startTime) : "";
  const initEndDate = route.params ? new Date(route.params.endTime) : "";

  const [startDate, setStartDate] = useState(initStartDate);
  const [endDate, setEndDate] = useState(initEndDate);
  const [isNap, setIsNap] = useState(false);
  const [err, setError] = useState({ startDate: false, endDate: false });

  async function submit() {
    if (!isValidInput(startDate, endDate, setError)) {
      console.log(err);
      return;
    }
    try {
      await SleepService.addSleepEntry(
        Math.floor(startDate.getTime() / 1000),
        Math.floor(endDate.getTime() / 1000),
        isNap
      );
      navigation.popToTop();
    } catch (error) {
      console.log(error);
    }
  }

  function displayErrorMessage(error) {
    if (error.startDate) {
      return <ErrorMessage message="Start date is invalid." />;
    }
    if (error.endDate) {
      return (
        <ErrorMessage message="End date is invalid. Please input a wake time past the sleep time." />
      );
    }
  }

  function toggleSwitch() {
    setIsNap((previousState) => !previousState);
  }

  return (
    <Screen style={styles.container}>
      <ScrollView style={{ padding: 10 }} keyboardShouldPersistTaps="handled">
        <View
          style={{
            backgroundColor: colors.white,
            padding: 10,
            borderColor: "black",
            borderRadius: 10,
            borderWidth: 1,
            flexDirection: "row",
            alignItems: "center",
            marginVertical: 10,
            justifyContent: "space-between",
          }}
        >
          <AppText style={[styles.text, { fontWeight: "bold" }]}>
            Is this a nap?
          </AppText>
          <Switch
            trackColor={{ false: "#767577", true: "#81b0ff" }}
            thumbColor={isNap ? "#f5dd4b" : "#f4f3f4"}
            ios_backgroundColor="#3e3e3e"
            onValueChange={toggleSwitch}
            value={isNap}
          />
        </View>
        <View style={{ padding: 10 }}>
          <AppText style={styles.text}>
            What time did you start sleeping?
          </AppText>
          <AppDateTimePicker
            placeholder="Select a time"
            init={startDate}
            onChange={(date) => {
              setStartDate(date);
              setError({ startDate: false, endDate: false });
            }}
          />
          <AppText style={styles.text}>What time did you wake up?</AppText>
          <AppDateTimePicker
            placeholder="Select a time"
            init={endDate}
            onChange={(date) => {
              setEndDate(date);
              setError({ startDate: false, endDate: false });
            }}
          />
          {displayErrorMessage(err)}
          <TextButton
            style={styles.button}
            name="Submit"
            onPress={async () => {
              await submit();
            }}
          />
        </View>
      </ScrollView>
    </Screen>
  );
}

const styles = StyleSheet.create({
  button: {
    marginVertical: 10,
  },
  container: {
    flex: 1,
    padding: 10,
    justifyContent: "center",
    backgroundColor: colors.background,
  },
  text: {
    fontSize: 16,
  },
});

export default SleepEntryFormScreen;
