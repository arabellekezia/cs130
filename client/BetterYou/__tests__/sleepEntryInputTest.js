import {render, cleanup, fireEvent} from '@testing-library/react-native';
import React from 'react';
import SleepEntryFormScreen from "../app/Screens/SleepEntryFormScreen"

describe('Sleep entry input form', () => {
  let wrapper;
  beforeEach(() => {
      wrapper = render(<SleepEntryFormScreen route={{}} />);
    });
  afterEach(() => {
      cleanup();
      wrapper = null;
    });
  it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
  });

  it("Shows error message if no category is selected", () => {
    const {getByText, queryByTestId} = wrapper;
    
    const submitButton = getByText("Submit");
    fireEvent.press(submitButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    expect(validationError.props.children)
        .toBe("Start date is invalid.");
  });
});

