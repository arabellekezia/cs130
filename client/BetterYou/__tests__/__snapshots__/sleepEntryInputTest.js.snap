// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sleep entry input form renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "center",
        "padding": 10,
      },
      Object {
        "paddingTop": 44,
      },
    ]
  }
>
  <RCTScrollView
    collapsable={false}
    keyboardShouldPersistTaps="handled"
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      Object {
        "padding": 10,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "black",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 10,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Array [
              Object {},
              Array [
                Object {
                  "fontSize": 16,
                },
                Object {
                  "fontWeight": "bold",
                },
              ],
            ]
          }
        >
          Is this a nap?
        </Text>
        <RCTSwitch
          accessibilityRole="switch"
          collapsable={false}
          onChange={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onTintColor="#81b0ff"
          style={
            Array [
              Object {
                "height": 31,
                "width": 51,
              },
              Object {
                "backgroundColor": "#3e3e3e",
                "borderRadius": 16,
              },
            ]
          }
          thumbTintColor="#f4f3f4"
          tintColor="#767577"
          value={false}
        />
      </View>
      <View
        style={
          Object {
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Array [
              Object {},
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          What time did you start sleeping?
        </Text>
        <RCTSafeAreaView
          emulateUnlessSupported={true}
          style={
            Array [
              undefined,
              Object {
                "paddingTop": 44,
              },
            ]
          }
        >
          <RNGestureHandlerButton
            collapsable={false}
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
          >
            <View
              accessible={true}
              style={
                Object {
                  "backgroundColor": "#efefef",
                  "borderRadius": 10,
                  "flexDirection": "row",
                  "marginVertical": 10,
                  "padding": 15,
                  "width": "100%",
                }
              }
            >
              <Text />
              <Text
                style={
                  Array [
                    Object {},
                    Object {
                      "color": "#7e7e7e",
                      "flex": 1,
                      "fontSize": 18,
                    },
                  ]
                }
              >
                Select a time
              </Text>
              <Text />
            </View>
          </RNGestureHandlerButton>
          <Modal
            animationType="none"
            hardwareAccelerated={false}
            isVisible={false}
            onHide={[Function]}
            transparent={true}
            visible={false}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "black",
                  "bottom": 0,
                  "height": 1334,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 750,
                }
              }
            />
          </Modal>
          <Modal
            animationType="none"
            hardwareAccelerated={false}
            isVisible={false}
            onHide={[Function]}
            transparent={true}
            visible={false}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "black",
                  "bottom": 0,
                  "height": 1334,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 750,
                }
              }
            />
          </Modal>
        </RCTSafeAreaView>
        <Text
          style={
            Array [
              Object {},
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          What time did you wake up?
        </Text>
        <RCTSafeAreaView
          emulateUnlessSupported={true}
          style={
            Array [
              undefined,
              Object {
                "paddingTop": 44,
              },
            ]
          }
        >
          <RNGestureHandlerButton
            collapsable={false}
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
          >
            <View
              accessible={true}
              style={
                Object {
                  "backgroundColor": "#efefef",
                  "borderRadius": 10,
                  "flexDirection": "row",
                  "marginVertical": 10,
                  "padding": 15,
                  "width": "100%",
                }
              }
            >
              <Text />
              <Text
                style={
                  Array [
                    Object {},
                    Object {
                      "color": "#7e7e7e",
                      "flex": 1,
                      "fontSize": 18,
                    },
                  ]
                }
              >
                Select a time
              </Text>
              <Text />
            </View>
          </RNGestureHandlerButton>
          <Modal
            animationType="none"
            hardwareAccelerated={false}
            isVisible={false}
            onHide={[Function]}
            transparent={true}
            visible={false}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "black",
                  "bottom": 0,
                  "height": 1334,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 750,
                }
              }
            />
          </Modal>
          <Modal
            animationType="none"
            hardwareAccelerated={false}
            isVisible={false}
            onHide={[Function]}
            transparent={true}
            visible={false}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "black",
                  "bottom": 0,
                  "height": 1334,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 750,
                }
              }
            />
          </Modal>
        </RCTSafeAreaView>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#007AFF",
                "borderRadius": 10,
                "justifyContent": "center",
                "marginVertical": 10,
                "minWidth": 110,
                "paddingHorizontal": 30,
                "paddingVertical": 12,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {},
                  Object {
                    "color": "white",
                    "fontSize": 18,
                    "fontWeight": "600",
                  },
                ]
              }
            >
              Submit
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
